### DISCUSSION TITLE
[Question] Conceptual direction

### DISCUSSION THREAD

**Discussion by philippk80** [roc-streaming/roc-toolkit#131] [Discuss]
Hello,

I'm looking into this for a while now and I have some question about the path this project is going.
Which aim you want to achieve?
Is this a solution which becomes something like a realtime multiroom audio solution for endusers?
Or will this be something like the dante protocol?
Or will this be something like a toolset to achieve such aims as developer?

Thanks.

Best regards,

Phil

**Comment by gavv** [comment]
Hi, thanks for the interest!

The primary goal is a library for developers that abstracts network transport and allows to write real-time samples to one end and read them from another end, hiding low-level details like maintaining latency, adjusting clock rate, handling duplicated and lost packets, etc.

There are also command line tools allowing to do the same. They're mostly intended for developers or power users that prefer manual configuration.

Another important goal is fully working PulseAudio integration, which is an example of an end-user solution, but not the only possible one. This is a good reality check for the project.

I'm not familiar with Dante. Conceptually, it looks like Roc and Dante have some properties in common, however, there are notable differences:

* Roc is a software-only solution working on application layer instead of network layer
* Roc is not a protocol but an implementation, it relies on standard protocols like RTP and FECFRAME
* one of the primary goals of Roc is an acceptable quality of service on 802.11
* there are plans to add support for compressed audio too (at least Opus)

**Comment by philippk80** [comment]
Thanks for this fast answer!

This means that Roc will handle End to end transmission for me.
Will it handle synchronization above multiple devices too or will this be part of the developer who implements Roc?
Are there any tests which latencys are realistic over LAN?
This would be great to get an idea about Roc and what it is able to do.

**Comment by gavv** [comment]
> Will it handle synchronization above multiple devices too or will this be part of the developer who implements Roc?

Multiroom synchronization is surely a suitable feature for Roc, however, it is not implemented or planned yet.

Off the top of my head, we have some basis that will help with this:

* multicast should either work or can be implemented easily (didn't test it yet)
* we're already able to maintain fixed (to some degree) latency on receiver by adjusting the clock rate

However, currently there is no guarantee that two receivers would start exactly at the same time and would adjust the clocks synchronously enough.

That is an interesting idea, and it requires some research and testing.

> Are there any tests which latencys are realistic over LAN?

In general, it depends on many factors, in particular on the LAN type (Ethernet or Wifi), the access point type (in case of WiFi), the network load, the interference level, etc.

Some Wifi networks I saw didn't allow latencies lower than one or even two seconds. I'm using a 200-300ms latency with my access point at home. Glitches still occur from time to time (say, once per ten or twenty minutes) because the whole packet stream is sometimes delayed too much for some reason.

Ethernet, of course, allows a much lower latency.

Also note that ALSA/PulseAudio may add from tens to hundreds of milliseconds, depending on your hardware.

These numbers are very rough thought and require much more testing. Also, I don't yet understand the reason of the large delays on my access point (one hypothesis is that they're caused by the link layer retries).

I'm planning a more serious testing after the 0.1 release.

**Comment by philippk80** [comment]
Sounds great! 
I'm looking into a realtime solution for multiroom setups.
My idea is to even handle the TV Audio over network. This requires a latency lower than 30ms to be usable. So I have to proof which latency I can achieve. Maybe over an own network if needed. Wifi isn't really an requirement by me. You already pointed the problems by choosing this type of transmission.

Synchronization isn't needed in my case because one than more device in a room isn't one of my concepts of planning a multimedia system. And I think over multiple rooms the small difference between devices won't be that critical if all is set up with LAN.

This project sounds promising for me. I'm a full time developer myself. Not that deep into such low level developing but familiar with .net and angular development. So with some work I'm able to implement your stuff and make usable solutions for my usecase.

Thanks for the amount of informations.
I will watching this further. Maybe I can event contribute if I'm deeper into low level languages.

**Comment by philippk80** [comment]
200-300ms over wifi sounds promising by the way.

**Comment by gavv** [comment]
You're welcome.

BTW, there were plans to add bindings for higher level languages, probably Python and Go. There were no plans for .NET bindings but contributions are always welcome.

**Comment by philippk80** [comment]
For .NET binding there should be Windows support to make sense first. Maybe I will try to make it build on windows. 
A .NET binding would be a task that is interesting for me. Wanted to do that anyway as learning process.
Besides that I'm studying in my free time at an university. C++ is a module of that course. Maybe this would be a great opportunity to learn that stuff with fun and a real use case. This gives a big motivation boost. I need to step out of the limited .NET world. 😄

**Comment by gavv** [comment]
> For .NET binding there should be Windows support to make sense first. Maybe I will try to make it build on windows.

Great, but note that porting to Windows would require pretty much work: at least porting SCons files and reimplementing classes and functions under all "target_gnu", "target_posix", and "target_posixtime" directories.

Windows port is not a priority currently, but I could help with porting after 0.1 or 0.2 release.

**Comment by philippk80** [comment]
I will open an issue for that so everybody is able to share efforts about this.
I think all other would over complicate this issue. 
The general answers are given. I will close this issue.
